### ISSUE TITLE
torch.compile produces inconsistent outputs for stateful SNN model with custom LIF neurons

### ISSUE THREAD

**Issue by Blooming-Tree** [pytorch/pytorch#188074]
### 🐛 Describe the bug

Description

When applying `torch.compile` to a Spiking Neural Network (SNN) model with custom LIF (Leaky Integrate-and-Fire) neurons, the compiled model produces outputs that significantly deviate from eager execution, exceeding acceptable error thresholds.

output:
```
Baseline - Float64 Eager:

tensor([[1.0078e-03, 2.8890e-03, 5.2671e-05, 2.4415e-01, 3.5602e-05, 1.7735e-02,
         8.7766e-03, 5.5573e-04, 2.0581e-01, 3.1647e-02],
        [1.5078e-01, 1.6141e-03, 1.0401e-04, 3.7304e-01, 7.5636e-05, 5.3631e-02,
         2.3807e-03, 3.1563e-04, 1.3192e-01, 5.2723e-02]])

Float32 Compiled:
tensor([[2.1976e-03, 2.8693e-03, 9.2174e-05, 2.4533e-01, 1.4005e-05, 6.3263e-03,
         5.2553e-02, 1.9321e-04, 2.9438e-01, 2.4956e-03],
        [2.4401e-01, 3.3686e-04, 1.1013e-04, 3.8893e-01, 8.3838e-05, 2.4826e-02,
         6.7275e-03, 1.5189e-03, 2.2342e-01, 1.3097e-01]])
```
code:
```python
import os
import copy
import torch
import torch.nn as nn
import torch.nn.functional as F

class LIFNeuron(nn.Module):

    def __init__(self, threshold=1.0, decay=0.9):
        super().__init__()
        self.threshold = threshold
        self.decay = nn.Parameter(torch.tensor(decay), requires_grad=True)
        self.register_buffer('mem_potential', None)

    def reset(self):
        self.mem_potential = None

    def forward(self, x):
        if self.mem_potential is None:
            self.mem_potential = torch.zeros_like(x)
        self.mem_potential = self.decay * self.mem_potential + x
        spike = torch.sigmoid(10 * (self.mem_potential - self.threshold))
        self.mem_potential = self.mem_potential * (1 - spike)
        return spike

class SNN(nn.Module):

    def __init__(self, input_dim=784, hidden_dim=128, output_dim=10, time_steps=10):
        super().__init__()
        self.time_steps = time_steps
        self.fc1 = nn.Linear(input_dim, hidden_dim)
        self.lif1 = LIFNeuron(threshold=0.5, decay=0.8)
        self.fc2 = nn.Linear(hidden_dim, output_dim)
        self.lif2 = LIFNeuron(threshold=0.5, decay=0.8)

    def forward(self, x):
        batch_size = x.size(0)
        self.lif1.reset()
        self.lif2.reset()
        total_output = torch.zeros(batch_size, self.fc2.out_features, device=x.device)
        for _ in range(self.time_steps):
            spike_input = torch.bernoulli(x).detach() + x - x.detach()
            h = F.relu(self.fc1(spike_input))
            h = self.lif1(h)
            out = getattr(self, 'fc2')(h)
            spikes = self.lif2(out)
            total_output += spikes
        return total_output / self.time_steps

def get_default_model():
    input_dim = 784
    hidden_dim = 128
    output_dim = 10
    time_steps = 10
    model = SNN(input_dim=input_dim, hidden_dim=hidden_dim, output_dim=output_dim, time_steps=time_steps)
    return model

def get_sample_inputs():
    batch_size = 2
    input_dim = 784
    x = torch.rand(batch_size, input_dim)
    return (x,)


def main():
    torch.use_deterministic_algorithms(True)
    torch.set_float32_matmul_precision('highest')
    os.environ['CUBLAS_WORKSPACE_CONFIG'] = ':4096:8'
    torch._inductor.config.fallback_random = True
    torch.manual_seed(42)
    model = get_default_model()
    model = model.double()
    model.eval()
    inputs = get_sample_inputs()
    inputs = tuple(t.double() if isinstance(t, torch.Tensor) and t.is_floating_point() else t for t in inputs)
    with torch.no_grad():
        eager_output = model(*inputs)
    print(f'eager float64 output: {eager_output}')
    torch.manual_seed(42)
    model_fp32 = copy.deepcopy(model)
    model_fp32 = model_fp32.float()
    model_fp32.eval()
    inputs_fp32 = tuple(t.float() if isinstance(t, torch.Tensor) and t.is_floating_point() else t for t in inputs)
    compiled_model = torch.compile(model_fp32, mode="reduce-overhead")
    with torch.no_grad():
        compiled_output = compiled_model(*inputs_fp32)
    print(f'compile float32 output: {compiled_output}')
    abs_diff = (eager_output - compiled_output.double()).abs()
    rel_diff = abs_diff / (eager_output.abs() + 1e-8)
    is_inconsistent = (abs_diff > 0.01) | (rel_diff > 0.001)
    print(f'Abs diff max: {abs_diff.max().item():.6e}, mean: {abs_diff.mean().item():.6e}')
    print(f'Rel diff max: {rel_diff.max().item():.6e}, mean: {rel_diff.mean().item():.6e}')
    print(f'Inconsistent: {is_inconsistent.sum().item()} / {eager_output.numel()}')
if __name__ == '__main__':
    main()
```

### Error logs

eager float64 output: tensor([[1.0078e-03, 2.8890e-03, 5.2671e-05, 2.4415e-01, 3.5602e-05, 1.7735e-02,
         8.7766e-03, 5.5573e-04, 2.0581e-01, 3.1647e-02],
        [1.5078e-01, 1.6141e-03, 1.0401e-04, 3.7304e-01, 7.5636e-05, 5.3631e-02,
         2.3807e-03, 3.1563e-04, 1.3192e-01, 5.2723e-02]])
compile float32 output: tensor([[2.1976e-03, 2.8693e-03, 9.2174e-05, 2.4533e-01, 1.4005e-05, 6.3263e-03,
         5.2553e-02, 1.9321e-04, 2.9438e-01, 2.4956e-03],
        [2.4401e-01, 3.3686e-04, 1.1013e-04, 3.8893e-01, 8.3838e-05, 2.4826e-02,
         6.7275e-03, 1.5189e-03, 2.2342e-01, 1.3097e-01]])
Abs diff max: 9.323436e-02, mean: 2.451178e-02
Rel diff max: 4.987887e+00, mean: 1.007789e+00
Inconsistent: 20 / 20

 Inconsistency detected! This may be a torch.compile bug.

### Versions

PyTorch version: 2.10.0+cu126

OS: Ubuntu 24.04.1 LTS

GPU: NVIDIA GeForce RTX 4060 Laptop GPU

CUDA version: 12.6

Python version: 3.12.3

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @kurtamohler @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @jataylo @azahed98

**Comment by Sanskaaar1** [comment]
The original repro compares fp64 eager vs fp32 compiled (different dtypes), but there is a real bug underneath.

`aten.bernoulli.default` is registered in `core_aten_decompositions()` and decomposes into `torch.rand() + comparison`. This decomposition fires even when `fallback_random=True`. On CUDA, the ATen `rand` and `bernoulli` kernels consume Philox RNG state in different thread-to-element orders, same seed but different sequence.

Every other random op (`rand_like`, `randn_like`, `randint_like`, `Dropout`) matches eager correctly under `fallback_random=True`. Only `bernoulli` breaks.

## Why

`aten.bernoulli_` and `aten.bernoulli.p` are already in `extra_random_decomps` and correctly excluded when `fallback_random=True`. `aten.bernoulli.default` being in `core_aten_decompositions` instead looks like an oversight.

## Repro

```python
import torch
torch._inductor.config.fallback_random = True

x = torch.full((1, 10), 0.5, device="cuda")

torch.manual_seed(42)
eager = torch.bernoulli(x)

compiled = torch.compile(lambda x: torch.bernoulli(x))
torch.manual_seed(42)
result = compiled(x)

print(torch.equal(eager, result))  # False, should be True
```

## Proposed Fix

Prevent the decomposition when `fallback_random=True` and route to ATen via `FallbackKernel` + layout constraints. Two options:

1. **Move the decomposition to `extra_random_decomps`** follows existing pattern for `bernoulli_`/`bernoulli.p`, but removes it from `core_aten_decompositions()` 
2. **Filter in `select_decomp_table()`** adds a per-op special case

Happy to put up a PR, is there a preferred approach?

**Comment by malfet** [comment]
Closing as won't-fix: this is expected behavior, not a `torch.compile` correctness bug.

The repro compares an **eager float64** run against a **compiled float32** run, and draws random numbers inside `forward`. Two independent things make those draws differ, neither related to compilation:

- **The RNG sequence depends on dtype.** `torch.bernoulli` on a float64 tensor and on a float32 tensor produce completely different draws from the same seed on the same device, because float64 generation consumes two 32-bit words per element. On 2.13.0, `manual_seed(42)` followed by `torch.bernoulli(torch.full((1000,), 0.5, dtype=...))` mismatches in 517/1000 positions between the two dtypes, diverging at the very first draw. Your `spike_input = torch.bernoulli(x)` therefore samples a different spike train in the float64 and float32 paths regardless of `torch.compile`.
- **The two paths reach `forward` at different positions in the RNG stream.** The eager forward runs after model construction and `get_sample_inputs()` have drawn from the generator; the compiled forward runs immediately after a bare `manual_seed(42)` and a `copy.deepcopy` (which draws nothing).

`torch._inductor.config.fallback_random = True` aligns RNG *within* compiled regions; it cannot make two differently-typed or differently-positioned draws agree. Identical results are not guaranteed here — see the [Reproducibility](https://docs.pytorch.org/docs/stable/notes/randomness.html) note: "Completely reproducible results are not guaranteed across PyTorch releases, individual commits, or different platforms."

Concretely: **removing `torch.compile` from your script entirely** and comparing eager float64 against eager float32 (same model, same `torch.rand(2, 784)` input, same harness) still flags every output element by your own criterion:

```
EAGER float64 vs EAGER float32 - torch.compile NOT involved
  Abs diff max: 9.272455e-02   mean: 2.479650e-02
  Rel diff max: 5.851454e+00
  'Inconsistent' by the report's own criterion: 20 / 20
```

So the reported divergence is produced by the float64-vs-float32 comparison, not by the compiler.

To isolate `torch.compile`, run both paths in the same dtype with the same weights, and reset the seed immediately before each forward. #188425 was closed for the same reason.
